src/components/interface/cbuf_mgr/cos_alloc.c

```diff
@@ -0,0 +1,536 @@
+  if (unlikely(size<sizeof(__alloc_t))) goto err_out;
+
+  if (size<=__MAX_SMALL_SIZE) {
+    need=GET_SIZE(size);
+    ptr=__small_malloc(need);
+  } else {
+    need=PAGE_ALIGN(size);
+    cbuf_t cbid;
+    ptr = need ? do_cbuf_alloc(need, &cbid) : MAP_FAILED;
+    ptr->cbuf_id = cbid;
+  }
+  if (ptr==MAP_FAILED) goto err_out;
+  ptr->size=need;
+  return BLOCK_RET(ptr);
+err_out:
+  printc("\treturning malloc error\n");
+  return 0;
+}
+
+void* malloc(size_t size) __attribute__((alias("_alloc_libc_malloc")));
+
+void *__libc_calloc(size_t nmemb, size_t _size)
+{
+	size_t tot = nmemb*_size;
+	char *ret = malloc(tot);
+	
+	memset(ret, 0, tot);
+	return ret;
+}
+
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+/* gabep1 additions for allocations of pages. */
+
+void *alloc_page(void)
+{
+	struct free_page *fp;
+	void *a;
+
+	fp = page_list.next;
+	if (NULL == fp) {
+        cbuf_t cbid;
+		a = do_cbuf_alloc(PAGE_SIZE, &cbid);
+        // what if this fails?
+	} else {
+		page_list.next = fp->next;
+		fp->next = NULL;
+		a = (void*)fp;
+	}
+	
+	return a;
+}
+
+void free_page(void *ptr)
+{
+	struct free_page *fp;
+	
+	fp = (struct free_page *)ptr;
+	fp->next = page_list.next;
+	page_list.next = fp;
+
+	return;
+}
+
+#ifdef NIL
+
+void* __libc_calloc(size_t nmemb, size_t _size);
+void* __libc_calloc(size_t nmemb, size_t _size) {
+  register size_t size=_size*nmemb;
+  if (nmemb && size/nmemb!=_size) {
+    (*__errno_location())=ENOMEM;
+    return 0;
+  }
+  return malloc(size);
+}
+void* calloc(size_t nmemb, size_t _size) __attribute__((weak,alias("__libc_calloc")));
+
+void* __libc_realloc(void* ptr, size_t _size);
+void* __libc_realloc(void* ptr, size_t _size) {
+  register size_t size=_size;
+  if (ptr) {
+    if (size) {
+      __alloc_t* tmp=BLOCK_START(ptr);
+      size+=sizeof(__alloc_t);
+      if (size<sizeof(__alloc_t)) goto retzero;
+      size=(size<=__MAX_SMALL_SIZE)?GET_SIZE(size):PAGE_ALIGN(size);
+      if (tmp->size!=size) {
+	if ((tmp->size<=__MAX_SMALL_SIZE)) {
+	  void *new=_alloc_libc_malloc(_size);
+	  if (new) {
+	    register __alloc_t* foo=BLOCK_START(new);
+	    size=foo->size;
+	    if (size>tmp->size) size=tmp->size;
+	    if (size) memcpy(new,ptr,size-sizeof(__alloc_t));
+	    _alloc_libc_free(ptr);
+	  }
+	  ptr=new;
+	}
+	else {
+	  register __alloc_t* foo;
+	  size=PAGE_ALIGN(size);
+	  foo=mremap(tmp,tmp->size,size,MREMAP_MAYMOVE);
+	  if (foo==MAP_FAILED) {
+retzero:
+	    (*__errno_location())=ENOMEM;
+	    ptr=0;
+	  }
+	  else {
+	    foo->size=size;
+	    ptr=BLOCK_RET(foo);
+	  }
+	}
+      }
+    }
+    else { /* size==0 */
+      _alloc_libc_free(ptr);
+      ptr = NULL;
+    }
+  }
+  else { /* ptr==0 */
+    if (size) {
+      ptr=_alloc_libc_malloc(size);
+    }
+  }
+  return ptr;
+}
+void* realloc(void* ptr, size_t size) __attribute__((weak,alias("__libc_realloc")));
+
+#endif
+
+/********************* testing code on unix ***********************/
```

Comment:
You don't really support any of this Linux, user-level testing, so you might as well remove it.